### PR TITLE
Feature/detox integration

### PR DESCRIPTION
**Pull request by william2958**
#### Changes
* added Detox testing for both local and CI. Currently only iOS detox testing is working locally.

### CODE REVIEW

**Comment by florianlenz** [comment]
it seem's like the build is aborted since it takes 60 min's to run it.

File: bitrise.yml [hunk 1]
```diff
@@ -0,0 +1,437 @@
+format_version: "4"
```
Comment (florianlenz) [inline bitrise.yml]: This file need to be removed from the VSC. It contain's password's which should never be committed.

File: package.json [hunk 2]
```diff
@@ -27,9 +31,9 @@
     "events": "^1.1.1",
     "fetch-everywhere": "^1.0.5",
     "hoist-non-react-statics": "^2.3.1",
-    "i18next": "^10.3.0",
+    "i18next": "^10.6.0",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 3]
```diff
@@ -27,9 +31,9 @@
...
     "lodash": "^4.17.4",
-    "native-base": "^2.3.5",
+    "native-base": "^2.3.10",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 4]
```diff
@@ -42,21 +46,21 @@
     "react-native-image-crop-picker": "^0.19.1",
     "react-native-keyboard-aware-scroll-view": "^0.4.3",
     "react-native-modal": "^4.1.1",
-    "react-native-modal-dropdown": "^0.6.0",
-    "react-native-navigation": "^1.1.319",
+    "react-native-modal-dropdown": "^0.6.2",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 5]
```diff
@@ -42,21 +46,21 @@
...
+    "react-native-navigation": "^1.1.425",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 6]
```diff
@@ -42,21 +46,21 @@
...
     "react-native-qrcode": "^0.2.6",
     "react-native-qrcode-scanner": "0.0.29",
     "react-native-randombytes": "^3.0.0",
     "react-native-responsive": "^1.0.2",
     "react-native-segmented-control-tab": "^3.2.1",
-    "react-native-sensitive-info": "^5.1.0",
+    "react-native-sensitive-info": "^5.2.0",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 7]
```diff
@@ -42,21 +46,21 @@
...
     "react-native-simple-dialogs": "^0.3.1",
     "react-native-swiper": "^1.5.13",
     "react-native-vector-icons": "^4.4.2",
-    "react-navigation": "^1.0.0-beta.22",
+    "react-navigation": "^1.5.8",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 8]
```diff
@@ -42,21 +46,21 @@
...
     "react-redux": "^5.0.6",
     "readable-stream": "^1.0.33",
-    "realm": "^2.3.1",
+    "realm": "^2.3.3",
```
Comment (florianlenz) [inline package.json]: Why was this dependency changed?

File: package.json [hunk 9]
```diff
@@ -6,9 +6,13 @@
     "start": "node node_modules/react-native/local-cli/cli.js start",
     "lint": "eslint src __tests__ __mocks__",
     "lint:fix": "eslint --fix src __tests__ __mocks__",
-    "test": "flow && jest",
+    "test": "flow && jest __tests__",
```
Comment (florianlenz) [inline package.json]: Why exactly was the `__test__` added?

File: package.json [hunk 10]
```diff
@@ -6,9 +6,13 @@
...
+    "test:e2e:debug": "detox test --configuration ios.sim.debug --cleanup",
+    "test:e2e:debug:build": "detox build --configuration ios.sim.debug",
+    "test:e2e:release": "detox test --configuration ios.sim.release --cleanup",
+    "test:e2e:release:build": "detox build --configuration ios.sim.release",
     "test:update": "jest -u",
-    "test:coverage": "flow && jest --coverage --forceExit",
+    "test:coverage": "flow && jest __tests__ --coverage --forceExit",
```
Comment (florianlenz) [inline package.json]: Why exactly was the `__test__` added?

File: package.json (hunk 9 above)
Comment (william2958) [reply package.json]: Without the __test__ parameter it would try and run the __e2e__ folder as well which would trigger the detox tests and would fail

File: package.json (hunk 10 above)
Comment (william2958) [reply package.json]: See above

File: bitrise.yml (hunk 1 above)
Comment (william2958) [reply bitrise.yml]: Removed.

File: android/gradle.properties [hunk 11]
```diff
@@ -18,3 +18,8 @@
 # org.gradle.parallel=true
 
 android.useDeprecatedNdk=true
+android.enableAapt2=false
+MYAPP_RELEASE_STORE_FILE=my-release-key.keystore
+MYAPP_RELEASE_KEY_ALIAS=my-key-alias
```
[the remark this answers is not in the document]
Comment (william2958) [reply android/gradle.properties]: Removed all the variables. They were for testing and I forgot to get rid of them.

File: android/app/build.gradle [hunk 12]
```diff
@@ -101,14 +101,27 @@ android {
...
+        testBuildType System.getProperty('testBuildType', 'debug')  //this will later be used to control the test apk build type
+        missingDimensionStrategy "minReactNative", "minReactNative46" //read note
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+    signingConfigs {
+        release {
+            if (project.hasProperty('MYAPP_RELEASE_STORE_FILE')) {
+                storeFile file(System.getenv("HOME") + '/bitnation_android_keys_dev')
+                storePassword System.getenv("ANDROID_KEY_STORE_PASSWORD")
+                keyAlias System.getenv("BITRISEIO_ANDROID_KEYSTORE_ALIAS")
```
[the remark this answers is not in the document]
Comment (william2958) [reply android/app/build.gradle]: changed.

File: bitrise.yml (hunk 1 above)
Comment (seland) [reply bitrise.yml]: Looks like still not removed

File: e2e/firstTest.spec.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+describe('Example', () => {
```
Comment (seland) [inline e2e/firstTest.spec.js]: Is that file necessary? Looks like a dummy file that won't be run, since we run tests from `__e2e__`, but not `e2e`

File: e2e/init.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+const detox = require('detox');
```
Comment (seland) [inline e2e/init.js]: Do we use that init file or one inside `__e2e__` folder?

File: e2e/firstTest.spec.js (hunk 13 above)
Comment (william2958) [reply e2e/firstTest.spec.js]: Yes I have removed the e2e folder completely. It was not being used.

File: e2e/init.js (hunk 14 above)
Comment (william2958) [reply e2e/init.js]: see above

File: bitrise.yml (hunk 1 above)
Comment (william2958) [reply bitrise.yml]: Ok I have fixed it

Comment (florianlenz) [reply bitrise.yml]: k. I already changed the exposed env var's. Keep in mind that everything committed and pushed stay's forever in git.

File: android/app/build.gradle [hunk 15]
```diff
@@ -101,14 +101,25 @@ android {
     buildToolsVersion '25.0.1'
```
Comment (florianlenz) [inline android/app/build.gradle]: @seland can you please have a very close look at the changes? I don't like to change the gradle files.

File: android/gradle.properties [hunk 16]
```diff
@@ -18,3 +18,5 @@
 # org.gradle.parallel=true
 
 android.useDeprecatedNdk=true
+android.enableAapt2=false
```
Comment (florianlenz) [inline android/gradle.properties]: @william2958 I think you can remove this form the file. I am also not sure if it should be commited @seland?

File: android/gradle.properties [hunk 17]
```diff
@@ -18,3 +18,5 @@
...
+MYAPP_RELEASE_STORE_FILE=my-release-key.keystore
```
Comment (florianlenz) [inline android/gradle.properties]: @william2958 I think you can remove this form the file. I am also not sure if it should be commited @seland?

Comment (william2958) [reply android/gradle.properties]: The enableAapt2 line is a setup line that is specified by the detox get started guide, I am pretty sure this is required to build successfully. The other line however was just me trying to get the CI to pass, it is a meaningless property. It is basically dummy data. I am not sure if this file was to be not committed though

File: android/app/build.gradle (hunk 15 above)
Comment (florianlenz) [reply android/app/build.gradle]: @seland ping.

File: android/app/build.gradle [hunk 18]
```diff
@@ -101,14 +101,25 @@ android {
...
             abiFilters "armeabi-v7a", "x86"
         }
         vectorDrawables.useSupportLibrary = true
+        testBuildType System.getProperty('testBuildType', 'debug')  //this will later be used to control the test apk build type
+        missingDimensionStrategy "minReactNative", "minReactNative46" //read note
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
+    signingConfigs {
+        release {
+            storeFile file(System.getenv("HOME") + '/bitnation_android_keys_dev')
```
Comment (florianlenz) [inline android/app/build.gradle]: @william2958 can you please make it just an Env var that point's to the key store file? The name of the var could be `BITNATION_ANDROID_KEY_STORE`

File: android/gradle.properties (hunk 16 above)
Comment (william2958) [reply android/gradle.properties]: This change was in the setup-guide for Detox. It's part of the config
